ppx/query.mll

```diff
@@ -0,0 +1,113 @@
+{
+open Ppx_mysql_runtime.Stdlib
+
+type param =
+  { typ : string
+  ; opt : bool
+  ; name : string
+  ; of_string : string * string
+  ; to_string : string * string }
+
+type parsed_query =
+  { query : string
+  ; in_params : param list
+  ; out_params : param list }
+
+type parse_error =
+  [ `Bad_identifier of string
+  | `Unknown_type_spec of string
+  | `Unterminated_string
+  | `Escape_at_end ]
+
+let build_param spec opt name =
+  let open Result in
+  begin match spec with
+  | "int" -> Ok ("int", ("Pervasives", "int_of_string"), ("Pervasives", "string_of_int"))
+  | "int32" -> Ok ("int32", ("Int32", "of_string"), ("Int32", "to_string"))
+  | "int64" -> Ok ("int64", ("Int64", "of_string"), ("Int64", "to_string"))
+  | "string" -> Ok ("string", ("Ppx_mysql_runtime", "identity"), ("Ppx_mysql_runtime", "identity"))
+  | spec -> Error (`Unknown_type_spec spec)
+  end >>= fun (typ, of_string, to_string) ->
+  Ok {typ; opt = (opt = "?"); name; of_string; to_string}
+}
+
+let escape = '\\'
+let squot = '\''
+let dquot = '"'
+let quot = squot | dquot
+let digit = ['0'-'9']
+let lower = ['a'-'z']
+let upper = ['A'-'Z']
+let underscore = '_'
+let ident = (lower | underscore) (lower | upper | underscore | digit)*
+let spec = (lower | upper | underscore | digit)+
+
+rule main_parser buf acc_in acc_out = parse
+  | quot as delim
+    {Buffer.add_char buf delim;
+    quotation_parser buf acc_in acc_out delim lexbuf}
+  | '%' (spec as spec) ('?'? as opt) '{' ([^ '}' ]+ as name) '}'
```

Comment:
I think it would be nice if this could make sure the `name` will end up being a valid OCaml identifier. I think it might be nicer than with `check_ident`, but I am open to discussion.

Reply:
My first implementation was as you suggested. But then I realised that if the `name` was not a valid OCaml identifier, then the lexer would of course simply transcribe that part of the query verbatim without raising any errors, when in practice if someone writes something like `SELECT id FROM users WHERE id = %int{ID}` we'd want the ppx to report that this is a mistake instead of accepting the query verbatim (and failing at runtime).